rust-sdk/whirlpool/src/decrease_liquidity.rs

```diff
@@ -0,0 +1,481 @@
+
+    let position_token_account_address = get_associated_token_address_with_program_id(
+        &authority,
+        &position_mint_address,
+        &position_mint_info.owner,
+    );
+    let lower_tick_array_address =
+        get_tick_array_address(&position.whirlpool, lower_tick_array_start_index)?.0;
+    let upper_tick_array_address =
+        get_tick_array_address(&position.whirlpool, upper_tick_array_start_index)?.0;
+
+    let tick_array_infos =
+        rpc.get_multiple_accounts(&[lower_tick_array_address, upper_tick_array_address])?;
+
+    let lower_tick_array_info = tick_array_infos[0]
+        .as_ref()
+        .ok_or("Lower tick array info not found")?;
+    let lower_tick_array = TickArray::from_bytes(&lower_tick_array_info.data)?;
+    let lower_tick = &lower_tick_array.ticks[get_tick_index_in_array(
+        position.tick_lower_index,
+        lower_tick_array_start_index,
+        pool.tick_spacing,
+    )? as usize];
+
+    let upper_tick_array_info = tick_array_infos[1]
+        .as_ref()
+        .ok_or("Upper tick array info not found")?;
+    let upper_tick_array = TickArray::from_bytes(&upper_tick_array_info.data)?;
+    let upper_tick = &upper_tick_array.ticks[get_tick_index_in_array(
+        position.tick_upper_index,
+        upper_tick_array_start_index,
+        pool.tick_spacing,
+    )? as usize];
+
+    let fees_quote = collect_fees_quote(
+        pool.clone().into(),
+        position.clone().into(),
+        lower_tick.clone().into(),
+        upper_tick.clone().into(),
+        transfer_fee_a,
+        transfer_fee_b,
+    )?;
+
+    let unix_timestamp = SystemTime::now().duration_since(UNIX_EPOCH)?.as_secs();
+    let rewards_quote = collect_rewards_quote(
+        pool.clone().into(),
+        position.clone().into(),
+        lower_tick.clone().into(),
+        upper_tick.clone().into(),
+        unix_timestamp,
+        get_current_transfer_fee(reward_infos[0].as_ref(), current_epoch),
+        get_current_transfer_fee(reward_infos[1].as_ref(), current_epoch),
+        get_current_transfer_fee(reward_infos[2].as_ref(), current_epoch),
+    )?;
+
+    let mut required_mints: Vec<TokenAccountStrategy> = Vec::new();
+
+    if quote.liquidity_delta > 0 || fees_quote.fee_owed_a > 0 || fees_quote.fee_owed_b > 0 {
+        required_mints.push(TokenAccountStrategy::WithoutBalance(pool.token_mint_a));
+        required_mints.push(TokenAccountStrategy::WithoutBalance(pool.token_mint_b));
+    }
+
+    for i in 0..3 {
+        if rewards_quote.rewards[i].rewards_owed > 0 {
+            required_mints.push(TokenAccountStrategy::WithoutBalance(
+                pool.reward_infos[i].mint,
+            ));
+        }
+    }
+
+    let token_accounts = prepare_token_accounts_instructions(rpc, authority, required_mints)?;
+
+    let mut instructions: Vec<Instruction> = Vec::new();
+    instructions.extend(token_accounts.create_instructions);
+
+    if position.liquidity > 0 {
```

Comment:
`UpdateFeesAndRewards` is not needed in close context.
If position.liquidity is not zero, `DecreaseLiquidityV2` will be executed and it has same effect of `UpdateFeesAndRewards`.
(TS SDK is same)

Reply:
Yep I think TS puts them in a SET first to get the list of uniques